.github/workflows/tests.yml

```diff
@@ -42,26 +45,78 @@ jobs:
         uses: actions/setup-node@v1
         with:
           node-version: ${{ matrix.node-version }}
-      - name: yarn install, build and test
-        run: |
-          yarn install --immutable
-          yarn lint
-          yarn build
-          yarn test:ci
+      - name: Get yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=dir::$(yarn config get cacheFolder)"
+      - uses: actions/cache@v2
+        id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-${{ matrix.node-version }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-${{ matrix.node-version }}-yarn-
+      - name: yarn install
+        run: yarn install --immutable
+      - name: yarn build
+        run: yarn build
+      - name: zip build
+        run: zip -r build.zip .
+      - name: upload build
+        uses: actions/upload-artifact@v2
+        with:
+          name: build-${{ matrix.node-version }}
+          path: build.zip
+
+
+  lint:
+    runs-on: ubuntu-latest
+    needs: build
+
+    steps:
+      - uses: actions/download-artifact@v2
+        with:
+          name: build-12.x
+      - name: unzip
+        run: unzip build.zip -d .
+      - name: lint
+        run: yarn lint
+        env:
+          CI: true
+
+  test:
+    runs-on: ubuntu-latest
+    needs: build
+
+    strategy:
+      matrix:
+        node-version: [10.x, 12.x]
+
+    steps:
+      - uses: actions/download-artifact@v2
+        with:
+          name: build-${{ matrix.node-version }}
+      - name: unzip
+        run: unzip build.zip -d .
+      - name: unit tests
+        run: yarn test:ci
         env:
           CI: true
 
   integration_test:
     runs-on: ubuntu-latest
 
-    needs: check_skip
+    needs: build
 
     strategy:
       matrix:
         image: ['latest-develop', 'latest-rc']
 
     steps:
-      - uses: actions/checkout@v1
+      - uses: actions/download-artifact@v2
+        with:
+          name: build-12.x
+      - name: unzip
+        run: unzip build.zip -d .
```

Comment:
these steps could maybe be abstracted away. A custom action maybe?

Reply:
there are zip/unzip actions out there, just felt like its overkill to replace a single command

Reply:
could just help with porting to different os

Reply:
we could use this https://github.com/TonyBogdanov/zip ; it uses the exact same tooling but launches an alpine linux container to do so :D

Reply:
no it's fine. as we talked directly, let's not overoptimize it